reco_utils/common/tf_utils.py

```diff
@@ -49,15 +57,122 @@ def pandas_input_fn(
+    seed=None,
+    queue_capacity=1000,
+    num_threads=1
+):
+    """Numpy input function for TensorFlow high-level API Estimator.
+    
+    This function is exactly the same as tensorflow's (tf.estimator.inputs.numpy_input_fn),
+    except this one allows the seed for shuffle.
+        
+    Args:
+        x (dict): Dictionary of feature (numpy array) object.
+        y (np.array): Labels. `None` if absent.
+        batch_size (int): Size of batches to return.
+        num_epochs (int): Number of epochs to iterate over data. If `None` will run forever.
+        shuffle (bool): If True shuffles the queue. Avoid shuffle at prediction time.
+        seed (int): Random seed for shuffle.
+        queue_capacity (int): Size of queue to accumulate.
+        num_threads (int): Number of threads used for reading and enqueueing.
+            In order to get deterministic results, use `num_threads=1`.
+        
+    Returns:
+        Function that has signature of ()->(dict of features, targets)
+    """
+    def input_fn():
+        # Note that `x` should not be used after conversion to ordered_dict_data,
+        # as type could be either dict or array.
+        ordered_dict_data = _validate_and_convert_features(x)
+
+        # Deep copy keys which is a view in python 3
+        feature_keys = list(ordered_dict_data.keys())
+
+        if y is None:
+            target_keys = None
+        elif isinstance(y, dict):
+            if not y:
+                raise ValueError('y cannot be empty dict, use None instead.')
+
+            ordered_dict_y = collections.OrderedDict(
+                sorted(y.items(), key=lambda t: t[0]))
+            target_keys = list(ordered_dict_y.keys())
+
+            duplicate_keys = set(feature_keys).intersection(set(target_keys))
+            if duplicate_keys:
+                raise ValueError('{} duplicate keys are found in both x and y: '
+                                 '{}'.format(len(duplicate_keys), duplicate_keys))
+
+            ordered_dict_data.update(ordered_dict_y)
+        else:
+            target_keys = _get_unique_target_key(ordered_dict_data)
+            ordered_dict_data[target_keys] = y
+
+        if len(set(v.shape[0] for v in ordered_dict_data.values())) != 1:
+            shape_dict_of_x = {k: ordered_dict_data[k].shape for k in feature_keys}
+
+            if target_keys is None:
+                shape_of_y = None
+            elif isinstance(target_keys, str):
+                shape_of_y = y.shape
+            else:
+                shape_of_y = {k: ordered_dict_data[k].shape for k in target_keys}
+
+            raise ValueError('Length of tensors in x and y is mismatched. All '
+                             'elements in x and y must have the same length.\n'
+                             'Shapes in x: {}\n'
+                             'Shapes in y: {}\n'.format(shape_dict_of_x, shape_of_y))
+
+        queue = feeding_functions._enqueue_data(  # pylint: disable=protected-access
```

Comment:
is this doing some kind of prefetching?

Reply:
Yes, but with the new approach I replied to your previous comment, which is simply adding a wrapper for `tf.estimator.inputs.numpy_input_fn` to support shuffle w/ seed, we can hide all of this code. Let me know if that make sense

Reply:
cool